Implement transport client

rosstimothy opened the pull request:
Provides a client that can be used to connect to and interact with the transport service in `lib/srv/transport`. The client abstracts the fact that a gRPC stream is being used for the `net.Conn` created by `DialCluster` and `DialHost`.

This also moves `lib/utils/grpc/stream` to `api/utils/grpc/stream` so that the client and server can make use of the same stream abstractions.

Part of #19812

strideynet commented on api/client/proxy/transport/transportv1/client.go:
```diff
@@ -0,0 +1,228 @@
+		for {
+			req, err := stream.Recv()
+			if err != nil {
+				if errors.Is(err, io.EOF) {
```
This err check treats an `io.EOF` the same as any other error - is there a missing behaviour here for an error that isn't an `io.EOF` ?

strideynet replied:
Do we want to consider warn/debug logging for these ?

strideynet commented on api/client/proxy/transport/transportv1/client_test.go:
```diff
@@ -0,0 +1,588 @@
+	// Start.
+	go func() {
+		if err := s.Serve(lis); err != nil {
+			panic(fmt.Sprintf("Serve returned err = %v", err))
```
Is there an advantage to using a panic instead of `t.FailNow` or similar here ?

strideynet replied:
I don't have particularly strong feelings (I'm not the world's biggest fan of `assert`/`require`) - but I do think that a large part of our test corpus currently uses `assert`/`require` and without any directive to suggest we are moving away from that, where it works, it's probably best to remain consistent with that.

strideynet reviewed (approved):
Really nice and neat PR.

It definitely would be nice once all the parts of this come together for us to consider putting together some Go benchmarks E2E for this - it's definitely code I can see performance regressions occurring within.

rosstimothy committed:
Implement transport client

Provides a client that can be used to connect to and interact with
the transport service in `lib/srv/transport`. The client abstracts
the fact that a gRPC stream is being used for the `net.Conn` created
by `DialCluster` and `DialHost`.

This also moves `lib/utils/grpc/stream` to `api/utils/grpc/stream`
so that the client and server can make use of the same stream
abstractions.

Part of #19812
